### PR TITLE
[dx11] add render passes and debugging

### PR DESCRIPTION
**Pull request by fkaa**
Re-do format conversion to be more comprehensive and less hacky

Add Bgra8 -> Buffer transfer

Add basic render passes (missing partial clear/resolves)

Add debugging using `ID3DUserDefinedAnnotation`. This functionality is behind a `#[cfg(debug_assertions)]` and is further restricted from being used on deferred contexts when RenderDoc is not used.

Separate `MemorySync` into Flush/Invalidate variants

r? @kvark 

Fixes #issue
PR checklist:
- [ ] `make` succeeds (on *nix)
- [ ] `make reftests` succeeds
- [ ] tested examples with the following backends:
- [ ] `rustfmt` run on changed code

### CODE REVIEW

File: src/backend/dx11/shaders/copy.hlsl [hunk 1]
```diff
@@ -116,6 +116,11 @@ uint Uint8x2ToUint16(uint2 data)
     return dot(min(data, 0xFF), 1 << uint2(0, 8));
 }
 
+uint4 Float4ToUint8x4(float4 data)
+{
+    return uint4(floor(data * 255 + .5f));
```
Comment (kvark) [inline src/backend/dx11/shaders/copy.hlsl]: does it need a `floor`?

File: src/backend/dx11/src/conv.rs [hunk 2]
```diff
@@ -225,6 +111,321 @@ pub fn map_format(format: Format) -> Option<DXGI_FORMAT> {
     Some(format)
 }
 
+#[derive(Debug, Clone)]
+pub struct DecomposedDxgiFormat {
+    pub typeless_format: DXGI_FORMAT,
```
Comment (kvark) [inline src/backend/dx11/src/conv.rs]: whoa, this is looking very thorough!
Perhaps, we can omit `_format` suffices in the fields?

File: src/backend/dx11/src/conv.rs [hunk 3]
```diff
@@ -225,6 +111,321 @@ pub fn map_format(format: Format) -> Option<DXGI_FORMAT> {
...
+    pub copy_uav_format: Option<DXGI_FORMAT>,
+    pub copy_srv_format: Option<DXGI_FORMAT>,
+}
+
+// TODO: we probably want to pass in usage flags or similar to allow for our `typeless_format`
+//       field to only contain the input format (eg. depth only rather than typeless likely
+//       improves perf since the driver doesn't need to expose internals)
+//
+// TODO: we also want aspect for determining depth/stencil
+pub fn decompose_format(format: DXGI_FORMAT) -> DecomposedDxgiFormat {
```
Comment (kvark) [inline src/backend/dx11/src/conv.rs]: let's make it a constructor funftion of the target struct

File: src/backend/dx11/src/debug.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+mod temp {
+    use winapi::shared::minwindef::{BOOL, INT}; 
+    use winapi::um::unknwnbase::{IUnknown, IUnknownVtbl}; 
+    use winapi::um::winnt::LPCWSTR; 
+
+    RIDL!{#[uuid(0xb2daad8b, 0x03d4, 0x4dbf, 0x95, 0xeb, 0x32, 0xab, 0x4b, 0x63, 0xd0, 0xab)]
+    interface ID3DUserDefinedAnnotation(ID3DUserDefinedAnnotationVtbl):
+        IUnknown(IUnknownVtbl) {
+        fn BeginEvent(
+            Name: LPCWSTR,
```
Comment (kvark) [inline src/backend/dx11/src/debug.rs]: any reason those are split into multiple lines?

File: src/backend/dx11/src/device.rs [hunk 5]
```diff
@@ -1341,21 +1348,44 @@ impl hal::Device<Backend> for Device {
...
+                        view_kind: image::ViewKind::D2,
+                        format: decomposed.dsv_format.unwrap(),
+                        range: image::SubresourceRange {
+                            aspects: format::Aspects::COLOR,
+                            levels: mip..(mip + 1),
+                            layers: layer..(layer + 1)
+                        }
+                    };
+
+                    depth_stencil_views.push(self.view_image_as_depth_stencil(&view).map_err(|_| device::BindError::WrongMemory)?);
```
Comment (kvark) [inline src/backend/dx11/src/device.rs]: nit: let's not stretch this line as long

File: src/backend/dx11/src/device.rs [hunk 6]
```diff
@@ -1685,6 +1715,9 @@ impl hal::Device<Backend> for Device {
                             pso::DescriptorType::StorageImage => {
                                 unsafe { *handle = Descriptor(image.uav_handle.clone().unwrap().as_raw() as *mut _); }
                             },
+							pso::DescriptorType::InputAttachment => {
```
Comment (kvark) [inline src/backend/dx11/src/device.rs]: indentation is broken here

File: src/backend/dx11/src/device.rs [hunk 7]
```diff
@@ -1614,7 +1644,7 @@ impl hal::Device<Backend> for Device {
                 StorageTexelBuffer | StorageBuffer |
                 InputAttachment | StorageBufferDynamic |
                 StorageImage => {
-                    binding.handle_offset = num_s + num_t + num_u + u;
+                    binding.handle_offset = num_s + num_t + num_c + u;
```
Comment (kvark) [inline src/backend/dx11/src/device.rs]: generally, a small fix like this would go as a separate commit, to have this one strictly as a feature commit
I don't mind mixing it at this point, when things are still in flux

File: src/backend/dx11/src/device.rs [hunk 8]
```diff
@@ -1764,12 +1797,15 @@ impl hal::Device<Backend> for Device {
...
     {
+        let _scope = debug::DebugScope::with_name(&self.context, &format!("FlushMappedRanges"));
 
         // go through every range we wrote to
         for range in ranges.into_iter() {
             let &(memory, ref range) = range.borrow();
             let range = memory.resolve(range);
 
+
+            let _scope = debug::DebugScope::with_name(&self.context, &format!("Range({:?})", range));
```
Comment (kvark) [inline src/backend/dx11/src/device.rs]: Unfortunate thing is that the string is constructed even if debug scopes are disabled. Sounds like the creation of a scope needs to be a macro itself

File: src/backend/dx11/src/device.rs [hunk 9]
```diff
@@ -2035,11 +2085,10 @@ impl hal::Device<Backend> for Device {
                 format: config.color_format,
                 storage_flags: image::StorageFlags::empty(),
                 // NOTE: not the actual format of the backbuffer(s)
-                typed_raw_format: dxgiformat::DXGI_FORMAT_UNKNOWN,
-                dxgi_format: format,
+                decomposed_format: decomposed.clone(),
                 num_levels: 1,
                 num_mips: 1,
-                internal
+                internal: internal
```
Comment (kvark) [inline src/backend/dx11/src/device.rs]: uh?

File: src/backend/dx11/src/internal.rs [hunk 10]
```diff
@@ -591,7 +606,14 @@ impl Internal {
                     let info = copy.borrow();
                     self.update_buffer_image(context, &info);
 
-                    let uav = dst.get_uav(info.image_layers.level, 0).unwrap().as_raw();
+                    debug::debug_marker(context, &format!("{:?}", info));
+
+                    // TODO: multiple layers? do we introduce a stride and do multiple dispatch
```
Comment (kvark) [inline src/backend/dx11/src/internal.rs]: sounds reasonable

File: src/backend/dx11/src/lib.rs [hunk 11]
```diff
@@ -268,11 +270,11 @@ impl hal::Instance for Instance {
                         heap_index: 0,
                     },
                     hal::MemoryType {
-                        properties: Properties::CPU_VISIBLE,
+                        properties: Properties::CPU_VISIBLE | Properties::CPU_CACHED,
```
Comment (kvark) [inline src/backend/dx11/src/lib.rs]: we expose no uncached CPU memory any more?

File: src/backend/dx11/src/lib.rs [hunk 12]
```diff
@@ -723,6 +741,86 @@ impl hal::queue::RawCommandQueue<Backend> for CommandQueue {
 
 }
 
+#[derive(Derivative, Clone)]
+#[derivative(Debug)]
+struct AttachmentClear {
+    subpass_id: Option<pass::SubpassId>,
+    #[derivative(Debug="ignore")]
+    value: Option<command::ClearValueRaw>,
```
Comment (kvark) [inline src/backend/dx11/src/lib.rs]: I think having `Debug` implemented for `ClearValueRaw` (manually) would be handy for other backends and even users

File: src/backend/dx11/src/lib.rs [hunk 13]
```diff
@@ -997,19 +1100,50 @@ impl hal::command::RawCommandBuffer<Backend> for CommandBuffer {
...
+                            self.context.ClearRenderTargetView(
+                                image.get_rtv(level, layer).unwrap().as_raw(),
+                                &color.float32
+                            );
+                        }
+                    }
+                }
+            }
+
+            // TODO: clear foreach level, layer on DSVs
```
Comment (kvark) [inline src/backend/dx11/src/lib.rs]: aren't you iterating the levels/layers already?

File: src/backend/dx11/src/lib.rs [hunk 14]
```diff
@@ -1196,32 +1337,32 @@ impl hal::command::RawCommandBuffer<Backend> for CommandBuffer {
...
-                        host_memory: sync.host_ptr,
-                        sync_range: sync.range.clone(),
-
-                        buffer: sync.device_buffer
-                    });
+            {
+                let coherent_buffers = set.coherent_buffers.lock();
+                for sync in coherent_buffers.flush_coherent_buffers.borrow().iter() {
+                    // TODO: merge sync range if a flush already exists
+                    if self.flush_coherent_memory.iter().position(|m| m.buffer == sync.device_buffer).is_none() {
```
Comment (kvark) [inline src/backend/dx11/src/lib.rs]: `iter().any(...)`

File: src/backend/dx11/src/lib.rs [hunk 15]
```diff
@@ -1255,32 +1398,31 @@ impl hal::command::RawCommandBuffer<Backend> for CommandBuffer {
...
-                    self.flush_coherent_memory.push(MemorySync {
-                        working_buffer: None,
-                        working_buffer_size: 0,
-
-                        host_memory: sync.host_ptr,
-                        sync_range: sync.range.clone(),
-
-                        buffer: sync.device_buffer
-                    });
+            {
```
Comment (kvark) [inline src/backend/dx11/src/lib.rs]: is there a point in this scope now?

File: src/backend/dx11/src/lib.rs [hunk 16]
```diff
@@ -1454,9 +1596,23 @@ bitflags! {
...
+enum SyncRange {
+    Whole,
+    Partial(Range<u64>)
+}
+
+#[derive(Debug, Clone)]
+pub struct MemoryFlush {
+    host_memory: *mut u8,
+    sync_range: SyncRange,
+
```
Comment (kvark) [inline src/backend/dx11/src/lib.rs]: nit: why empty line?

File: src/backend/dx11/src/lib.rs [hunk 17]
```diff
@@ -1615,33 +1791,75 @@ impl Memory {
...
+                //      the whole range (TODO: allow for partial updates)
+                //
+                //   3. the general case, without any `UNIFORM` usage has no restrictions on
+                //      partial updates, so we upload the specified range
+                //
+                if buffer.usage.contains(Usage::UNIFORM) && buffer.usage != Usage::UNIFORM {
+                    MemoryFlush {
+                        host_memory: unsafe { ptr.offset(buffer_range.start as _) },
+                        sync_range: SyncRange::Whole,
+
```
Comment (kvark) [inline src/backend/dx11/src/lib.rs]: weird newlines

File: src/backend/dx11/src/lib.rs [hunk 18]
```diff
@@ -1691,9 +1909,31 @@ impl ::std::fmt::Debug for ShaderModule {
 unsafe impl Send for ShaderModule { }
 unsafe impl Sync for ShaderModule { }
 
-#[derive(Debug)]
-pub struct RenderPass;
-#[derive(Debug)]
+#[derive(Debug, Clone)]
+pub struct SubpassDesc {
+    pub(crate) color_attachments: Vec<pass::AttachmentRef>,
```
Comment (kvark) [inline src/backend/dx11/src/lib.rs]: why not `pub`?

File: src/backend/dx11/src/device.rs (hunk 8 above)
Comment (fkaa) [reply src/backend/dx11/src/device.rs]: I'd hope that the compiler could optimize it out, but haven't actually checkedi if that's the case. Do you mean a macro that simply does `#[cfg(...)]` on the function call?

File: src/backend/dx11/src/debug.rs (hunk 4 above)
Comment (fkaa) [reply src/backend/dx11/src/debug.rs]: Taken from winapi, which has some unorthodox formatting. Hopefully there'll be a new winapi release around next week

File: src/backend/dx11/src/lib.rs (hunk 12 above)
Comment (fkaa) [reply src/backend/dx11/src/lib.rs]: Agree. How would it handle [the union](https://github.com/gfx-rs/gfx/blob/master/src/hal/src/command/raw.rs#L45)? just print out all the variants?

File: src/backend/dx11/src/lib.rs (hunk 15 above)
Comment (fkaa) [reply src/backend/dx11/src/lib.rs]: shorter scope one the mutex lock

File: src/backend/dx11/src/device.rs (hunk 8 above)
Comment (kvark) [reply src/backend/dx11/src/device.rs]: I mean a macro like `debug_scope!("tralalal {:?}", something)`

File: src/backend/dx11/src/lib.rs (hunk 12 above)
Comment (kvark) [reply src/backend/dx11/src/lib.rs]: Just print out the `u32` one

File: src/backend/dx11/src/lib.rs (hunk 13 above)
Comment (fkaa) [reply src/backend/dx11/src/lib.rs]: yup, heat must be getting to me 😓

File: src/backend/dx11/src/lib.rs (hunk 11 above)
Comment (fkaa) [reply src/backend/dx11/src/lib.rs]: Required for WR, not to mention that reads are cached already since we copy over from gpu memory to our "mapped" Vec<u8>
